### PR TITLE
[Example] Add sagpool example for pytorch backend

### PR DESCRIPTION
**Pull request by lygztq**
## Description

Add sagpool (in paper [Self-Attention Graph Pooling](https://arxiv.org/pdf/1904.08082.pdf)) example for pytorch backend

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/sagpool/layer.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        self.score_layer = conv_op(in_dim, 1)
+        self.non_linearity = non_linearity
+    
+    def forward(self, graph:dgl.DGLGraph, feature:torch.Tensor):
+        score = self.score_layer(graph, feature).squeeze()
+        perm, next_batch_num_nodes = topk(score, self.ratio, get_batch_id(graph.batch_num_nodes()), graph.batch_num_nodes())
+        feature = feature[perm] * self.non_linearity(score[perm]).view(-1, 1)
+        graph = dgl.node_subgraph(graph, perm)
+
+        # node_subgraph currently does not support batch-graph,
```
Comment (yzh119) [inline examples/pytorch/sagpool/layer.py:34]: Is it related to #2422 ? I'll fix the behavior, thanks for letting us know.

**Review by yzh119 — APPROVED** [review]
Well done!

File: examples/pytorch/sagpool/layer.py (hunk 1 above)
Comment (lygztq) [reply examples/pytorch/sagpool/layer.py:34]: Yes, this is related to #2422 . I think this issue is common in graph classification task with pooling layers.

File: examples/pytorch/sagpool/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+```bash
+python grid_search.py --device ${your_device_id_here} --num_trials ${num_of_trials_here}
+```
+
+Performance
+-------------------------
+
+NOTE: We do not perform grid search or finetune here, so there may be a gap between results in paper and our results. Also, we only perform 10 trials for each experiment, which is different from 200 trials per experiment in the paper.
+
+**The global architecture result**
```
Comment (hetong007) [inline examples/pytorch/sagpool/README.md:73]: Is our variance number from 95% confidence interval? If so, I suggest we change to standard deviation to provide comparable information.

File: examples/pytorch/sagpool/README.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+- NumFeats: 14
+- NumClasses: 2
+
+
+How to run example files
+--------------------------------
+In the sagpool folder, run
+
+```bash
+python main.py --dataset ${your_dataset_name_here}
```
Comment (hetong007) [inline examples/pytorch/sagpool/README.md:54]: First, seems it is case-sensitive. Probably we can list all valid names for datasets.

Especially for `DD`, in the following tables it is named `D&D` but in the code it is `DD`.

File: examples/pytorch/sagpool/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import numpy as np
+
+
+def collate_fn(batch):
+    """
+    collate_fn for dataset batching
+    transform ndata to tensor (in gpu is available)
+    """
+    graphs, labels = map(list, zip(*batch))
+    #cuda = torch.cuda.is_available()
```
Comment (hetong007) [inline examples/pytorch/sagpool/dataloader.py:13]: Could we have it cleaned up?

File: examples/pytorch/sagpool/grid_search_config.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+    "arch": ["hierarchical", "global"],
+    "hidden": [16, 32, 64, 128],
+    "pool_ratio": [0.25, 0.5],
+    "lr": [1e-2, 5e-2, 1e-3, 5e-3, 1e-4, 5e-4],
+    "weight_decay": [1e-2, 1e-3, 1e-4, 1e-5],
+    "dataset": ["DD", "PROTEINS", "NCI1", "NCI109", "Mutagenicity"]
+}
```
Comment (hetong007) [inline examples/pytorch/sagpool/grid_search_config.json:8]: Missing an empty final line

File: examples/pytorch/sagpool/main.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        loss += F.nll_loss(out, batch_labels, reduction="sum").item()
+        correct += pred.eq(batch_labels).sum().item()
+    return correct / num_graphs, loss / num_graphs
+
+
+def main(args):
+    # Step 1: Prepare graph data and retrieve train/validation/test index ============================= #
+    dataset = LegacyTUDataset(args.dataset, raw_dir=args.dataset_path)
+
+    # add self loop. We add self loop for each graph here since the function "add_self_loop" does not
```
Comment (hetong007) [inline examples/pytorch/sagpool/main.py:117]: @yzh119 Do you think we can support batched graph transformation in a more principled way?

**Review by hetong007 — COMMENTED** [review]
Overall good work! Just a few comments from my side.

File: examples/pytorch/sagpool/README.md (hunk 2 above)
Comment (lygztq) [reply examples/pytorch/sagpool/README.md:73]: Yes, our variance number is from 95% conf interval, I will change the result with std later.

File: examples/pytorch/sagpool/README.md (hunk 3 above)
Comment (lygztq) [reply examples/pytorch/sagpool/README.md:54]: Users can get a list of valid dataset name from the help-string in argparser, I will also add it in README

File: examples/pytorch/sagpool/README.md [hunk 7]
```diff
@@ -80,17 +80,17 @@ NOTE: We do not perform grid search or finetune here, so there may be a gap betw
...
 **The hierarchical architecture result**
 | Dataset       | paper result (hierarchical)      | ours (hierarchical)         |
 | ------------- | -------------------------------- | --------------------------- |
-| D&D           | 76.45 (0.97)                     | 76.47 (4.45)                |
-| PROTEINS      | 71.86 (0.97)                     | 72.41 (1.89)                |
-| NCI1          | 67.45 (1.11)                     | 72.34 (1.69)                |
-| NCI109        | 67.86 (1.41)                     | 69.73 (1.41)                |
-| Mutagenicity  | N/A                              | 74.71 (2.74)                |
+| D&D           | 76.45 (0.97)                     | 75.97 (2.34)                |
+| PROTEINS      | 71.86 (0.97)                     | 70.54 (2.81)                |
```
Comment (hetong007) [inline examples/pytorch/sagpool/README.md:93]: Can you please double check? I see many std in this table being larger than the previous one, just to make sure that you have filled in the correct numbers.
